Wedge node

johnhaddon opened the pull request:
This adds a Wedge node, which makes it very simple to dispatch the same job multiple times with a range of values on the upstream nodes. Wedges are connected downstream from the job to be repeated, and create multiple copies of the upstream tasks, each with a unique value from a numeric range, or from a color range or from an explicit list of numbers or strings. This value is communicated upstream as a context variable where it can be used to influence any aspect of the job.

In the example below I've placed two wedge nodes in series below an Appleseed render job. One wedge defines the value for "metallicness" in the range 0-1, with a step of 0.25. The next does the same but for roughness. Upstream these variables are used in an expression to control the parameters of a Disney material. Because they're connected in series, when dispatched they generate 25 different renders with all permutations of those two variables. The outputs from these renders are then combined into a single image using the recently added ImageLoop node, to create the image you see in the viewer.

![wedge](https://cloud.githubusercontent.com/assets/1133871/8594915/9fb46272-263d-11e5-8d90-05508ff9310b.png)

dictoon commented:
Pretty cool!

Narann commented:
I love this! :+1:

andrewkaufman commented:
Looks like that test is still failing...

andrewkaufman commented:
I think we got a bit confused in our private discussion about how this should work. Now that I see what you've come up with, I think the "simple integer variable" that you thought I was suggesting was actually what I was considering the number of steps to take (minimum 2). So rather than having the user specify an explicit step value (e.g. 0.25) and having the only possible application of it be a linear stepping, I was suggesting that we specify numSteps (e.g. 4) and then also a step function (e.g. linear, exponential, logarithmic, ramp/spline), and that would all be contained within the FloatRange mode. So you'd be able to get some fancier value variation without having to switch to FloatList mode and compute them manually.

That's what we've provided for our wedge tools in other packages, and those extra step functions were explicitly requested by users in the past, though to be honest I don't know how often they actually use anything other than linear.

Anyways, I'm happy to go with this until a user requests more control, unless you think that approach I described makes sense to do now. I guess doing it later might be tricky without adding a new mode...

andrewkaufman commented:
Also, I think we will need that `ExecutableContextVariables` node for wedge blocking, since `ExecutableRender` hashes the whole context currently.

johnhaddon commented:
I'm not overly attached to the way I've done it, so happy to change to something better now rather than later. I kindof went with using the step for floats so that it matched the way the int range is specified, where numSteps wouldn't make sense due to integer rounding. But step for floats has it's own issues - a range of 0-1 with a step of 0.1 doesn't hit 1 due to float rounding errors. I was also wondering whether or not it might be nicer to use frame range syntax for ints (1-10x2 style), so step might not even make sense there. Although I suspect having start/stop/step in separate for plugs for ints does have benefits when wanting to drive a wedge setup with expressions.

I'm not familiar with the settings for our internal wedge tools so I'll ping you on chat later to hash something out.

ExecutableContextVariables sounds good, and I think we could slightly simplify the implementation of that and Wedge if we had an ExecutableContextProcessor base class. I noticed on another internal ticket that we might have a use for a FrameMask type node to limit execution of certain passes to a different frame range - that too could probably build on the same base class. I don't see any of that as being a lot of work (maybe half a day), so should I update this pull request to include it?

andrewkaufman commented:
Yep, sounds good

johnhaddon commented:
OK, I have :

- Updated the FloatRange mode to use a number-of-steps rather than a size-of-step approach. This fixes a problem with float precision meaning the max value would be missed unexpectedly, and also sets us up for fancy remappings in the future.
- Added a TaskContextVariables node. I felt this was a better user-facing name than ExecutableContextVariables - hopefully that is OK.
- Disabled the test on Travis, after confirming with 3delight support that the error messages indicate a basic problem running 3delight on the Travis VM, rather than any genuine problem in Gaffer or 3delight.
